csvdiffer replacement

milljm opened the pull request:
A working version of the new csvdiff executable.

See how far this gets through the pipline...

Commit:
A simple working csvdiff callable

We are not finished yet. Now we need to focus on using a costum comparision file (one generated from an exodiff dump).

Refs #12363

Commit:
Generate a summary

Refs #12363

Commit:
Parse a config file (wip)

Commit:
Obey spec params for rel_tol and abs_zero

Add ability for tests spec file to use a custom comparison file

Commit:
Add newline at EOF

Commit:
remove unnecessary variables

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/257240/) on 6b9b0d3 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/12487/site)

This comment will be updated on new commits.

Commit:
Stream line comparison file parser

Allow 'time' field to populate available fields to be compared with (or not)

Commit:
Fix bugs, spelling, etc

Return on error during table generation called from summary report.
Simplify the output when generating a comparison file.
Fix some spelling errors.

Commit:
Format diff message

Use a more pythonic way to format the diff message.
Assume file[0] is the 'good/gold' file.

Commit:
Indexes incorrect

Representing wrong value due to zero based index array.

Commit:
Remove variables no longer required

Commit:
Skip multi-csvdiff file test

This test now needs to be broken up into multiple tests in order for the new
csvdiff executable to function properly (like a multi-exodiff test would).

The current test uses features not possible to exist in a non-stateful executable
(if a field exists in any csv file tested... then ignore missing fields in others).

Closes #12363

permcody commented on framework/contrib/csvdiff/csvdiff:
```diff
@@ -0,0 +1,386 @@
+#!/usr/bin/env python2.7
```
I don't think that this script should be in contrib. Contrib is for external projects that we've copied into MOOSE. If you want to open source this as a project under your profile, then it would be appropriate for it to be a contrib. As it stands it's a utility that's just part of MOOSE. I might recommend that you just put it in the scripts directory (moose/scripts).

permcody commented on framework/contrib/csvdiff/csvdiff:
```diff
@@ -0,0 +1,386 @@
+                for param in set(a_line.split()).intersection(tollerance_params):
+                    custom_params['RELATIVE'] = re.findall(param + "\s+([0-9e.-]+)", a_line)[0]
+                for param in set(a_line.split()).intersection(zero_params):
+                    custom_params['ZERO'] = re.findall(param + "\s+([0-9e.-]+)", a_line)[0]
```
Minor: You use set(a_line_split()) 4 times here. That's worth creating a variable so you don't have to create the set over and over.

permcody replied:
Also you have the same regex pattern in two places. You might just create the regex object once and use it twice. Helps with code maintainability too.

permcody commented on framework/contrib/csvdiff/csvdiff:
```diff
@@ -0,0 +1,386 @@
+    def __init__(self, args):
+        CSVTools.__init__(self)
+        self.files = args.summary
+        self.abs_zero = float(str(args.abs_zero))
```
Convert to str then float? Why is that necessary?

permcody commented on framework/contrib/csvdiff/csvdiff:
```diff
@@ -0,0 +1,386 @@
+        self.files = args.csv_file
+        self.config = args.comparison_file
+        self.abs_zero = float(str(args.abs_zero))
+        self.rel_tol = float(str(args.relative_tolerance))
```
Again with the double conversion. I don't think this is right or necessary.

permcody reviewed (changes requested):
This is looking good. Just a few oddities in the code that should be reviewed before this goes in. That'll be good anyway since this hasn't gone through testing in a long time. We need a fresh set of testing before we can merge.

permcody commented:
FYI - After we get this in, I'm going to have you clean it up a bit. All of that override stuff will just completely go away. It's a mess and won't be necessary with the new comparison file stuff.

I'm testing the differ locally with some really nasty files. I just created a custom comp file and it seems to be working well - nice!

Commit:
Refactor a regex loop

remove type changes that were no longer needed

Commit:
Remove unused assigned variable

moosebuild commented:
Job [Private App tests PETSc alt](https://civet.inl.gov/job/272604/) on 0a81112 : invalidated by @permcody

Xgels server was down

permcody reviewed (approved):
This looks pretty good. I have a few more architecture things that I'd like to change, but I don't want to hold this up any longer. I'll open up separate tickets with specific tasks.

Thanks!

bwspenc commented:
A couple of comments on this:

1) If we're going to use something that looks like the exodiff file format for the cmp files, we should go all the way and actually use the same file format.  The only things we need to do are to change:
`default value tolerances:` to `GLOBAL VARIABLES`, and `default time step tolerances:` to `TIME STEPS`. I just tested exodiff, and it happily runs with just these two inputs (it doesn't need the coords and all that other stuff if it's just checking globals), so if we made this change, someone would be able to use the same .cmp file to check global variables interchangeably with both exodiff and csvdiff.  This makes it so someone can generate one of these cmp files with exodiff easily if they don't happen to know where csvdiff.py is located.

2) Do we really need to deprecate the existing options for specifying individual variable tolerances in the tests file? I think it's great that we have the option for a separate .cmp file for cases when there are a lot of variables that we need to tweak. However, we often have cases where there are just one or two variables to tweak, and I prefer listing those tolerances in the tests file (like we do for global tolerances) to creating separate .cmp files.  With exodiff, we often end up with lots of .cmp files kicking around just to tweak a couple of tolerances. I actually made an issue on SEACAS a while back to support overriding tolerances for individual variables at the command line hoping that we could provide that kind of behavior for exodiff: https://github.com/gsjaardema/seacas/issues/122. Does it really hurt the maintainability of the code that much to have two ways of specifying those tolerances? We're always going to have two ways of specifying the global tolerances.

permcody commented:
In response to Ben:

1. Yes! We definitely want the same format to the extent possible. BTW - Jason made a "-summary" flag for this so it _will_ generate a comparison file stub for you to work with. I agree with everything else you said.

2. You might be able to convince me on this. However, My issue is that the existing "override" options are a little screwy. They build up a set of columns seen over a set of files, they don't act on a single target. I was simply unable to get the existing options to work for my very simple case of screening out a few columns that I didn't want to check without further hacking the code. I'm open to tweaking tolerances for individual variables on the command line, but if you need to do something more complicated like look for a set of columns in different files, that kind of behavior is getting a little too complicated for command line options. Let's keep this utility small and sensible and not make it into too much of a Swiss Army knife.

bwspenc commented:
@permcody I'll look into this a little more to understand how those override options behave.  If there's something about the behavior of the old tool that doesn't make sense, I'm all for changing that.
